ClientRuntimes/Python/msrestazure/msrestazure/azure_active_directory.py

```diff
@@ -216,7 +228,70 @@ def clear_cached_token(self):
             raise_with_traceback(KeyError, "Unable to clear token.")
 
 
-class UserPassCredentials(OAuthTokenAuthentication, AADMixin):
+class AADRefreshMixin(object):
+    """
+    Additional token refresh logic
+    """
+
+    def refresh_session(self):
+        """Return updated session if token has expired, attempts to
+        refresh using newly acquired token.
+
+        :rtype: requests.Session.
+        """
+        if self.token.get('refresh_token'):
+            try:
+                return self.signed_session()
+            except Expired:
+                pass
+        self.set_token()
+        return self.signed_session()
+
+
+class AADTokenCredentials(AADMixin):
+    """
+    Credentials objects for AAD token retrieved through external process
+    e.g. Python ADAL lib.
+
+    Optional kwargs may include:
+    - china (bool): Configure auth for China-based service,
```

Comment:
Actually, everybody will have to write this all the time.
If you change the env in creds, this change your client automatically, so it's the best code to be "generic"
Because I don't see a way to relate ResourceClient and managementEndpointUrl automatically.
Could the Swagger have a new annotation or something to tag the host, to be replaced on runtime by the correct cloud Host?